### DISCUSSION TITLE
Dash AG Grid + DMC

### DISCUSSION THREAD

**Discussion by AnnMarieW** [snehilvj/dash-mantine-components#467] [Tips and Tricks] [answered]
# Cell Editors
How to use DMC V>=0.14 components as cell editors in Dash AG Grid

To use these features, place the JavaScript file containing the` AllFunctionalComponentEditors` function in the assets folder of your Dash project. You can then define the Mantine components in your app's column definitions (columnDefs), referencing `AllFunctionalComponentEditors `as the cellEditor. 

Note - this app is using the Mantine Dark theme.  The grid sets a dark theme using `className="ag-theme-alpine-dark"`  Note that the dmc components are correctly styled with the dark theme.    For a light theme, the grid's default of `ag-thime-alpine` works well with the default Mantine light theme.  

[See code in GitHub](https://github.com/snehilvj/dmc-docs/blob/main/help_center/aggrid/app_cell_editors.py)
[Live Demo on PyCafe](https://py.cafe/dash.mantine.components/dash-ag-grid-dmc-cell-editors)

![image](https://github.com/user-attachments/assets/de0b8d37-077f-4cca-bb6b-16242541193a)
![image](https://github.com/user-attachments/assets/b1329cb7-46ee-4d9b-86b4-27c4ceca241f)

**Comment by AnnMarieW** [comment]
# Cell Renderer with dmc.Button with DashIconify icons

This is the example from the the [Dash AG Grid Docs](https://dash.plotly.com/dash-ag-grid/cell-renderer-components#example-5:-cell-render-with-dmc.button-with-dashiconify-icons.) updated for dmc>=0.14.0

![image](https://github.com/user-attachments/assets/66b98835-46e6-4556-a807-59ef2c4f47c5)


```python
"""
styling with custom cell renderer

Note:
Custom components  must be defined in the dashAgGridComponentFunctions.js in assets folder.
"""

import json
import dash_ag_grid as dag
from dash import Dash, html, dcc, Input, Output, _dash_renderer
import pandas as pd
import dash_mantine_components as dmc
import dash_iconify

_dash_renderer._set_react_version("18.2.0")
data = {
    "ticker": ["AAPL", "MSFT", "AMZN", "GOOGL"],
    "company": ["Apple", "Microsoft", "Amazon", "Alphabet"],
    "price": [154.99, 268.65, 100.47, 96.75],
    "buy": ["Buy" for _ in range(4)],
    "sell": ["Sell" for _ in range(4)],
    "watch": ["Watch" for _ in range(4)],
}
df = pd.DataFrame(data)

columnDefs = [
    {
        "headerName": "Stock Ticker",
        "field": "ticker",
    },
    {"headerName": "Company", "field": "company", "filter": True},
    {
        "headerName": "Last Close Price",
        "type": "rightAligned",
        "field": "price",
        "valueFormatter": {"function": """d3.format("($,.2f")(params.value)"""},
    },
    {
        "field": "buy",
        "cellRenderer": "DMC_Button",
        "cellRendererParams": {
            "variant": "outline",
            "leftIcon": "ic:baseline-shopping-cart",
            "color": "green",
            "radius": "xl"
        },
    },
    {
        "field": "sell",
        "cellRenderer": "DMC_Button",
        "cellRendererParams": {
            "variant": "outline",
            "leftIcon": "ic:baseline-shopping-cart",
            "color": "red",
            "radius": "xl"
        },
    },
{
        "field": "watch",
        "cellRenderer": "DMC_Button",
        "cellRendererParams": {
            "rightIcon": "ph:eye",
        },
    },
]


defaultColDef = {
    "resizable": True,
    "sortable": True,
    "editable": False,
}


grid = dag.AgGrid(
    id="custom-component-dmc-btn-grid",
    columnDefs=columnDefs,
    rowData=df.to_dict("records"),
    columnSize="autoSize",
    defaultColDef=defaultColDef,
)


app = Dash(__name__)

app.layout = dmc.MantineProvider(html.Div(
    [
        dcc.Markdown("Example of cellRenderer with dash-mantine-components Button  and DashIconify icons"),
        grid,
        html.Div(id="custom-component-dmc-btn-value-changed"),
    ],
    style={"margin": 20},
))


@app.callback(
    Output("custom-component-dmc-btn-value-changed", "children"),
    Input("custom-component-dmc-btn-grid", "cellRendererData"),
)
def showChange(n):
    return json.dumps(n)


if __name__ == "__main__":
    app.run_server(debug=True)


"""
Put the following in the dashAgGridComponentFunctions.js file in the assets folder

---------------

var dagcomponentfuncs = window.dashAgGridComponentFunctions = window.dashAgGridComponentFunctions || {};

dagcomponentfuncs.DMC_Button = function (props) {
    const {setData, data} = props;

    function onClick() {
        setData();
    }
    let leftIcon, rightIcon;
    if (props.leftIcon) {
        leftIcon = React.createElement(window.dash_iconify.DashIconify, {
            icon: props.leftIcon,
        });
    }
    if (props.rightIcon) {
        rightIcon = React.createElement(window.dash_iconify.DashIconify, {
            icon: props.rightIcon,
        });
    }
    return React.createElement(
        window.dash_mantine_components.Button,
        {
            onClick,
            variant: props.variant,
            color: props.color,
            leftSection: leftIcon,
            rightSection: rightIcon,        
            radius: props.radius,
            style: {
                margin: props.margin,
                display: 'flex',
                justifyContent: 'center',
                alignItems: 'center',
            },
        },
        props.value
    );
};



"""

```

**Reply by kyoshizzz** [reply]
sorry i am new with dash-ag-grid, 
how do i disabled the button with params 
if the params.data.last close price is higher than $150 then disabled else enabled

**Reply by AnnMarieW** [reply]
Hi @kyoshizzz 

You can pass a variable for the max price at which to disable the button in the `cellRendererParams`  for example:

```python
"cellRendererParams": {
            "variant": "outline",
            "leftIcon": "ic:baseline-shopping-cart",
            "color": "green",
            "radius": "xl",
             "disabledPrice": 150.00
        },

```

Then in the JavaScript function, you can set the disabled prop on the button:

```js

  return React.createElement(
        window.dash_mantine_components.Button,
        {
            onClick,
            variant: props.variant,
            color: props.color,
            leftSection: leftIcon,
            rightSection: rightIcon,
            radius: props.radius,
            style: {
                margin: props.margin,
                display: 'flex',
                justifyContent: 'center',
                alignItems: 'center',
            },
            disabled: props.disabledPrice && props.data.price > props.disabledPrice
        },
        props.value
    );
```

Note - for debugging, before the return, you can add:
`console.log(props)`
Then you can inspect the props in the browser.  You can see that the last close price is at `props.data.price`

**Comment by AnnMarieW** [comment]
# Cell Renderer - dmc.Card 

![image](https://github.com/user-attachments/assets/1ec18697-f3c8-44ad-8800-aa1f6adb1b2d)


```python
"""
styling with custom cell renderer - Cards

Note:
Custom components  must be defined in the dashAgGridComponentFunctions.js in assets folder.
"""


import dash_ag_grid as dag
from dash import Dash, html, dcc, _dash_renderer
import pandas as pd
import dash_mantine_components as dmc

_dash_renderer._set_react_version("18.2.0")
data = {

    "stats": ["Revenue", "EBITDA", "Net Income"],
    "results": ['$13,120,000','$1,117,000', '$788,000'],
    "change": [10.1, -22.1, 18.6],
    'comments': ['', '', '']
}
df = pd.DataFrame(data)

columnDefs = [
    {
        "headerName": "",
        "cellRenderer": "DMC_Card",
    },
    {
        "field": "comments",
        "editable": True,
        "cellEditorPopup": True,
        "cellEditor": "agLargeTextCellEditor",
        "minWidth": 300,
    },
]

grid = dag.AgGrid(
    columnDefs=columnDefs,
    rowData=df.to_dict("records"),
    columnSize="autoSize",
    dashGridOptions={"rowHeight": 100}
)


app = Dash(__name__)

app.layout = dmc.MantineProvider(html.Div(
    [
        dcc.Markdown("Example of cellRenderer with dash-mantine-components Card"),
        grid,
    ],
    style={"margin": 20},
), forceColorScheme="light")



if __name__ == "__main__":
    app.run_server(debug=True)


"""
Put the following in the dashAgGridComponentFunctions.js file in the assets folder

---------------

var dagcomponentfuncs = window.dashAgGridComponentFunctions = window.dashAgGridComponentFunctions || {};



dagcomponentfuncs.DMC_Card = function (props) {
    return React.createElement(
        window.dash_mantine_components.Card,
        {
            withBorder:true
        },
        React.createElement(
            window.dash_mantine_components.Text,
            {c:"dimmed", tt: "uppercase", fw:700},
            props.data.stats
        ),
        React.createElement(
            window.dash_mantine_components.Text,
            {size: "lg"},
            props.data.results
        ),
        React.createElement(
            window.dash_mantine_components.Text,
            {size: "sm", c: (props.data.change > 0) ? "green" : "red"},
            props.data.change + "%"
        )
    );
};



"""

```

**Comment by benweinberg89** [comment]
It would nice if this tutorial was located on the DMC website. I'm always having to dig to find this post.

**Reply by AnnMarieW** [reply]
Yes - that's a great idea.  I almost forgot about this post myself!

**Comment by AnnMarieW** [comment]
@benweinberg89 
Check out the new section of the docs here:  https://www.dash-mantine-components.com/dash-ag-grid

Feedback is welcome.  I didn't want to duplicate info that is in the dash-docs, but is it clear?  Anything else need to be added (or edited?)

**Comment by StijnvanderLippe** [comment, marked as answer]
I am using dmc.Select components in grid cells to allow users to select one of the available options. When you open the Select element and click one of the options, the Select component stays active (it's popped out). See this example from the PyCafe link in the original post.
<img width="682" height="433" alt="image" src="https://github.com/user-attachments/assets/ed25b4af-7e2d-41f3-a551-8e3332d6a8b6" />

The value does not get "locked in" (data registered in rowData of the table) until the user clicks away from the Select component. It seems, however, you cannot click away outside the table. The only way you can get the popped out element to "lock in" its value is to select another cell in the table (but if this is an editable cell, you will start editing), or one of the column headers (but this will sort the column). This is not a nice user experience.

Is there any way to make it so the Select element "locks in" its data when selecting one of the options in the dropdown? So that the user does not have to do an additional click away?

UPDATE:
In order to get the above working, do the following:
1. Add `"stopEditingWhenCellsLoseFocus": True` to `dashGridOptions` so that clicking out of the table stops editing.
2. Add `classNames={"dropdown": "ag-custom-component-popup"}` to your `dmc.Select` component to make sure the values get registered when editing stops. Otherwise clicking out of the cell does not actually register the option you selected in the `dmc.Select` component. From https://www.ag-grid.com/javascript-data-grid/component-filter/#custom-filters-containing-a-popup-element

**Reply by AnnMarieW** [reply]
Hi @StijnvanderLippe 

Great answer!  I didn't know about adding the className - that's awesome!  I'll update the example in PyCafe

**Reply by AnnMarieW** [reply]
Just a quick follow up.  When I updated the PyCafe app, it worked if I added `dashGridOptions={"stopEditingWhenCellsLoseFocus": True},`  But if I also added the `classNames` it no longer updated when I clicked outside the grid.  Is that the same for you?  What version of the grid are you running?

**Reply by StijnvanderLippe** [reply]
It's working fine for me. On the PyCafe demo that's currently live, with _only_ `"stopEditingWhenCellsLoseFocus": True`:

https://github.com/user-attachments/assets/c6ee20a5-6dd1-4645-9769-a39bf2374db6

Values don't get registered when clicking away.

If I add `classNames={"dropdown": "ag-custom-component-popup"}`, so 
`'cellEditorParams': {'component': dmc.Select(data=["A", "B", "C"], classNames={"dropdown": "ag-custom-component-popup"})}`
or
`'cellEditorParams': {'component': dmc.TagsInput(data=["A", "B", "C"], classNames={"dropdown": "ag-custom-component-popup"})}` 
(I don't know which selector to use for the `DatePickerInput` unfortunately):

https://github.com/user-attachments/assets/4d21ef36-19cc-4351-bdd8-6cd4d8861c02

Now if you click away the value gets registered as desired.

**Reply by AnnMarieW** [reply]
Great!  It's working now.   I must have had some caching issue previously.

Thanks again for the good solution :trophy:
